src/source.cpp

```diff
@@ -79,6 +90,121 @@ unique_ptr<Source> Source::create(pugi::xml_node node)
+{
+  // Check for domains to reject from
+  if (check_for_node(node, "domain_type")) {
+    std::string domain_type = get_node_value(node, "domain_type");
+    if (domain_type == "cell") {
+      domain_type_ = DomainType::CELL;
+    } else if (domain_type == "material") {
+      domain_type_ = DomainType::MATERIAL;
+    } else if (domain_type == "universe") {
+      domain_type_ = DomainType::UNIVERSE;
+    } else {
+      fatal_error(
+        std::string("Unrecognized domain type for constraint: " + domain_type));
+    }
+
+    auto ids = get_node_array<int>(node, "domain_ids");
+    domain_ids_.insert(ids.begin(), ids.end());
+  }
+
+  if (check_for_node(node, "time_bounds")) {
+    auto ids = get_node_array<double>(node, "time_bounds");
+    if (ids.size() == 2)
+      time_bounds_ = std::make_pair(ids[0], ids[1]);
+  }
+  if (check_for_node(node, "energy_bounds")) {
+    auto ids = get_node_array<double>(node, "energy_bounds");
+    if (ids.size() == 2)
+      energy_bounds_ = std::make_pair(ids[0], ids[1]);
+  }
+
+  // Check for how to handle rejected particles
+  if (check_for_node(node, "rejection_strategy")) {
+    std::string rejection_strategy = get_node_value(node, "rejection_strategy");
+    if (rejection_strategy == "kill") {
+      rejection_strategy_ = RejectionStrategy::KILL;
+    } else if (rejection_strategy == "resample") {
+      rejection_strategy_ = RejectionStrategy::RESAMPLE;
+    } else {
+      fatal_error(std::string(
+        "Unrecognized strategy source rejection: " + rejection_strategy));
+    }
+  } else {
+    // Default to resample rejected particles
+    rejection_strategy_ = RejectionStrategy::RESAMPLE;
+  }
+}
+
+SourceSite Source::sample(uint64_t* seed) const
+{
+  bool found = false;
+  SourceSite site;
+
+  while (!found) {
+    // Sample a particle randomly from list
+    site = this->sample_no_rejection(seed);
+
+    found = satisfies_constraints(site);
+    if (!found && rejection_strategy_ == RejectionStrategy::KILL) {
+      // Accept particle regardless but set wgt=0 to trigger garbage collection
+      // I.e. kill this particle immediately and pick the next
+      found = true;
+      site.wgt = 0.0;
+    }
+  }
+  return site;
+}
+
+bool Source::satisfies_constraints(SourceSite& s) const
+{
+  return satisfies_spatial_constraints(s.r) &&
+         satisfies_energy_constraints(s.E) &&
+         satisfies_time_constraints(s.time);
+}
+
+bool Source::satisfies_energy_constraints(double E) const
+{
+  return E > energy_bounds_.first && E < energy_bounds_.second;
+}
+
+bool Source::satisfies_time_constraints(const double time) const
+{
+  return time > time_bounds_.first && time < time_bounds_.second;
+}
+
+bool Source::satisfies_spatial_constraints(Position r) const
+{
+  bool found = false;
+  GeometryState geom_state;
+  geom_state.r() = r;
+
+  // Reject particle if it's not in the geometry at all
+  if (!(found = exhaustive_find_cell(geom_state)))
```

Comment:
OK so... Was this meant to be an `==`? I had to check this to figure out what the behavior is supposed to be.

https://stackoverflow.com/questions/6807393/using-assignment-as-a-condition-expression

Because that's such a common mistake for people, I think that this (C++ exclusive!) assignment-as-an-expression might be better to avoid. I'd prefer we do `found = ...` then the condition check below.